### PR TITLE
always check file checksum and ensure that in-url hash matches file cont...

### PR DESCRIPTION
**Pull request by lloyd**
...ents. issue #31

### CODE REVIEW

File: lib/connect-cachify.js [hunk 1]
```diff
@@ -100,7 +102,17 @@ exports.setup = function (assets, options) {
...
         }
       }
-      if (exists === true) {
+
+      // determine actual current hash of the file, it's worth the disk
+      // read to ensure we never serve bogus resources and poison caches
+      // issue #31
+      var md5 = crypto.createHash('md5');
+      try {
+        md5.update(fs.readFileSync(true_path));
```
Comment (ozten) [inline lib/connect-cachify.js]: I think we should use `fs.readFile` and move calling `next` into this callback.

File: test/connect-cachify-test.js [hunk 2]
```diff
@@ -236,7 +235,23 @@ exports.setup = nodeunit.testCase({
...
+        mddlwr;
+    mddlwr = cachify.setup(
+        assets, {
+          root: '/tmp'
+    });
+    var before = req.url;
+    mddlwr(req, resp, function () {
+      test.ok(req.url == before);
+      test.done();
+    });
```
Comment (ozten) [inline test/connect-cachify-test.js:255]: I don't quite understand what this tests.

I expected to see a bad md5 hash and a 404.

**Comment by ozten** [comment]
Very nice and clean patch!

File: lib/connect-cachify.js (hunk 1 above)
Comment (seanmonstar) [inline lib/connect-cachify.js]: +9000

File: lib/connect-cachify.js [hunk 3]
```diff
@@ -158,7 +170,6 @@ var hashify = function (resource, hash) {
     var md5 = crypto.createHash('md5');
     try {
       var data = fs.readFileSync(filename);
```
Comment (seanmonstar) [inline lib/connect-cachify.js:175]: egads! another sync file read!

File: test/connect-cachify-test.js (hunk 2 above)
Comment (lloyd) [inline test/connect-cachify-test.js:255]: this tests that the middleware does not replace the URL when the md5 doesn't match.

I decided to let the request fall through to the normal 404 handling of the application.  I didn't want to be too intrusive.  I could add a X-Cachify-Failed header to the request.

I _will_ add documentation to the test to make it clear.

Is this sufficient?

File: lib/connect-cachify.js (hunk 1 above)
Comment (lloyd) [inline lib/connect-cachify.js]: ok, on it.  you guys are right.

File: lib/connect-cachify.js (hunk 3 above)
Comment (lloyd) [inline lib/connect-cachify.js:175]: I'm going for minimalism in this patch.  I agree that we could (and should) go through and improve blocking at the very least, and maybe ozten's ideas about pre-calculating checksums and filewatching could be useful too (though I'm worried about the complexity of that - specifically that we'd be doing a lot of epolling that we'd need to clean up for the process to shutdown cleanly)

File: lib/connect-cachify.js [hunk 4]
```diff
@@ -100,21 +102,34 @@ exports.setup = function (assets, options) {
...
-              if (resp.getHeader(header)) resp.removeHeader(header);
+
+      // determine actual current hash of the file, it's worth the disk
+      // read to ensure we never serve bogus resources and poison caches
+      // issue #31
+      fs.readFile(true_path, function(err, contents) {
+        // ignore error
+        var md5 = crypto.createHash('md5');
+        if (contents) md5.update(contents);
+        var actualHash = md5.digest('hex').slice(0, 10);
```
Comment (lloyd) [inline lib/connect-cachify.js:113]: NOTE: unit tests rely on generating md5 sums for the empty string even if the file doesn't exist.  I went for minimalism and left this alone.

File: test/connect-cachify-test.js (hunk 2 above)
Comment (ozten) [inline test/connect-cachify-test.js:255]: Yeppers.

**Comment by ozten** [comment]
r+

I'd love to async the second one, not a blocker.

I'll wait for seanmonstar's feedback before hitting merge.

**Comment by seanmonstar** [comment]
While I'd love to rid all blocking calls from requests that I can, it _is_ out of scope for this patch. Also, we've been running with it that way for a while, and no huge issues have shown.

I forgot the important process: measure first, cut second.

**Comment by ozten** [comment]
Travis failed node 0.8 and passed 0.10.

I asked it to restart.

**Comment by ozten** [comment]
[Green](https://travis-ci.org/mozilla/connect-cachify/builds/6643761)

**Pull request merged**